File: Python/394. Decode String.py
class Solution:
    def decodeString(self, s: str) -> str:
        num_stack = list()
        str_stack = list()
        curr = ""
        num = 0
        for c in s:
            if '0' <= c <= '9':
                num = num * 10 + int(c)
            elif c == '[':
                num_stack.append(num)
                str_stack.append(curr)
                curr = ""
                num = 0
            elif c == ']':
                tmp = curr
                curr = str_stack.pop()
                n = num_stack.pop()
                curr += tmp * n
            else:
                curr += c
        return curr

File: 2024/394. Decode String.cpp
class Solution {
public:
    string repeat(string s, int n) {
        string res;
        for (int i = 0; i < n; i++) {
            res += s;
        }
        return res;
    }
    string decodeString(string s) {
        stack<int> countStack;
        stack<string> strStack;
        int count = 0;
        string str;
        for (char c : s) {
            if (isdigit(c)) {
                count = count * 10 + c - '0';
            } else if (c == '[') {
                countStack.push(count);
                strStack.push(str);
                count = 0;
                str.clear();
            } else if (c == ']') {
                string tmp = strStack.top();
                strStack.pop();
                tmp = tmp + repeat(str, countStack.top());
                countStack.pop();
                str = tmp;
            } else {
                str += c;
            }
        }
        return str;
    }
};